py/objstr.c: Enforce unicode-safety when creating str from vstr.

jimmo opened the pull request:
Fixes #9044 and https://github.com/adafruit/circuitpython/issues/6752 (@jepler @RetiredWizard).

Rather than just ensuring that streams are unicode safe, this ensures that all usages of mp_obj_str_from_vstr are safe, as there are other ways to trigger this bug (see https://github.com/micropython/micropython/issues/9044#issuecomment-1218053443).

By changing the `*_from_vstr` API slightly, it's possible to do this in a way that overall reduces code size and comes in at -104 bytes on PYBV11.

I hope this isn't a huge breaking change for anyone writing user c modules (or for CircuitPython to merge) but it's a very straightforward find-and-replace to fix.

jimmo committed:
py/objstr.c: Split mp_obj_str_from_vstr into bytes/str versions.

Previously the desired output type was specified. Now make the type part
of the function name. Because this function is used in a few places this
saves code size due to smaller call-site.

This makes `mp_obj_new_str_type_from_vstr` a private function of objstr.c
(which is almost the only place where the output type isn't a compile-time
constant).

This saves ~140 bytes on PYBV11.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

dlech commented:
LGTM. Should we add a test, e.g. reading a file or memory stream with bad UTF-8 data?

dpgeorge commented on extmod/modure.c:
```diff
@@ -372,7 +372,11 @@ STATIC mp_obj_t re_sub_helper(size_t n_args, const mp_obj_t *args) {
+        return mp_obj_new_str_from_vstr(&vstr_return);
+    } else {
+        return mp_obj_new_bytes_from_vstr(&vstr_return);
+    }
```
Would be interesting to find a test case where doing a re match does raise a unicode error, and then check how CPython handles it.

Also, to reduce code size, maybe a helper like `mp_obj_new_str_type_from_vstr_checked(type, vstr)` could be added?

dpgeorge commented on extmod/modujson.c:
```diff
@@ -67,7 +67,7 @@ STATIC mp_obj_t mod_ujson_dump_helper(size_t n_args, const mp_obj_t *pos_args, m
         vstr_init_print(&vstr, 8, &print_ext.base);
         mp_obj_print_helper(&print_ext.base, pos_args[0], PRINT_JSON);
-        return mp_obj_new_str_from_vstr(&mp_type_str, &vstr);
+        return mp_obj_new_str_from_vstr(&vstr);
```
Does this need to validate utf8?  I would think that it's all already been validated as part of the dump process, or at least the dump should never produce invalid utf8.

dpgeorge commented on extmod/modujson.c:
```diff
@@ -103,7 +103,7 @@ STATIC mp_obj_t mod_ujson_dumps(mp_obj_t obj) {
     vstr_init_print(&vstr, 8, &print);
     mp_obj_print_helper(&print, obj, PRINT_JSON);
-    return mp_obj_new_str_from_vstr(&mp_type_str, &vstr);
+    return mp_obj_new_str_from_vstr(&vstr);
```
Does in need validating?

dpgeorge commented on extmod/vfs.c:
```diff
@@ -360,7 +360,7 @@ mp_obj_t mp_vfs_getcwd(void) {
         vstr_add_str(&vstr, cwd);
     }
-    return mp_obj_new_str_from_vstr(&mp_type_str, &vstr);
+    return mp_obj_new_str_from_vstr(&vstr);
```
Can a filesystem have invalid chars in it?  How do you get the cwd if it does?  How does CPython handle this?

dpgeorge commented on ports/unix/modusocket.c:
```diff
@@ -559,7 +559,7 @@ STATIC mp_obj_t mod_socket_inet_ntop(mp_obj_t family_in, mp_obj_t binaddr_in) {
     }
     vstr.len = strlen(vstr.buf);
-    return mp_obj_new_str_from_vstr(&mp_type_str, &vstr);
+    return mp_obj_new_str_from_vstr(&vstr);
```
Does this need validating?

dpgeorge commented on py/modbuiltins.c:
```diff
@@ -467,7 +467,7 @@ STATIC mp_obj_t mp_builtin_repr(mp_obj_t o_in) {
     vstr_init_print(&vstr, 16, &print);
     mp_obj_print_helper(&print, o_in, PRINT_REPR);
-    return mp_obj_new_str_from_vstr(&mp_type_str, &vstr);
+    return mp_obj_new_str_from_vstr(&vstr);
```
Surely the output of PRINT_REPR doesn't need validating?

dpgeorge commented on py/parsenum.c:
```diff
@@ -160,7 +160,7 @@ mp_obj_t mp_parse_num_integer(const char *restrict str_, size_t len, int base, m
         mp_str_print_quoted(&print, str_val_start, top - str_val_start, true);
         mp_obj_t exc = mp_obj_new_exception_arg1(&mp_type_ValueError,
-            mp_obj_new_str_from_vstr(&mp_type_str, &vstr));
+            mp_obj_new_str_from_vstr(&vstr));
```
Does this need validating?

dpgeorge commented:
What about calls to `mp_obj_new_str()` and `mp_obj_new_str_of_type()`?  Do they also need to validate?

jimmo committed:
py/objstr.c: Optimise mp_obj_new_str_from_vstr for known-safe strings.

The new `mp_obj_new_str_from_vstr_safe` can be used when you know you
already have a unicode-safe string.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jepler commented:
Hi, thanks for making this PR. I look forward to replacing the quick patch in CircuitPython with it down the road, especially if some flash space is freed in the process.

jimmo committed:
py/objstr.c: Always validate utf-8 for mp_obj_new_str.

All uses of this are either tiny strings or not-known-to-be-safe.

Update comments for mp_obj_new_str_copy and mp_obj_new_str_of_type.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tests/unicode: Add test for invalid utf-8 file contents.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
I've added an optimised path `mp_obj_new_str_from_vstr_safe` for the cases outlined above (and one more in persistentcode.c).

> Should we add a test, e.g. reading a file or memory stream with bad UTF-8 data?

Good idea. Done.

> What about calls to `mp_obj_new_str()` and `mp_obj_new_str_of_type()`? Do they also need to validate?

I think the the former should always check (all the uses I could find were small strings or untrusted). I have added a commit to do this.

`mp_obj_new_str_of_type` calls `mp_obj_new_str`.

The other two interesting ones are:

`mp_obj_new_str_copy` -- this is always used on safe strings
`mp_obj_new_str_via_qstr` -- this is also always used on safe strings

I have added/updated the comments for both to state this as a requirement.

---

This now comes in at -72 bytes on PYBV11.

dpgeorge commented on py/obj.h:
```diff
@@ -789,11 +789,11 @@ mp_obj_t mp_obj_new_int_from_uint(mp_uint_t value);
-mp_obj_t mp_obj_new_str_via_qstr(const char *data, size_t len);
-mp_obj_t mp_obj_new_str_from_vstr(vstr_t *vstr);
+mp_obj_t mp_obj_new_str(const char *data, size_t len); // will check utf-8 (raises UnicodeError)
+mp_obj_t mp_obj_new_str_via_qstr(const char *data, size_t len); // must be known to be utf-8 safe
```
"input data must be valid utf-8"

dpgeorge commented on extmod/modujson.c:
```diff
@@ -67,7 +67,7 @@ STATIC mp_obj_t mod_ujson_dump_helper(size_t n_args, const mp_obj_t *pos_args, m
         vstr_init_print(&vstr, 8, &print_ext.base);
         mp_obj_print_helper(&print_ext.base, pos_args[0], PRINT_JSON);
-        return mp_obj_new_str_from_vstr(&mp_type_str, &vstr);
+        return mp_obj_new_str_from_vstr_safe(&vstr);
```
`mp_obj_new_str_from_utf8_vstr`

jimmo committed:
py/objstr.c: Always ensure mp_obj_str_from_vstr is unicode-safe.

Now that we have `mp_obj_new_str_type_from_vstr` (private helper used by
objstr.c) split from the public API (`mp_obj_new_str_from_vstr`), we can
enforce a unicode check at the public API without incurring a performance
cost on the various objstr.c methods (which are already working on known
unicode-safe strings).

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
Updated.

dpgeorge commented on py/objstr.h:
```diff
@@ -88,8 +88,8 @@ mp_obj_t mp_obj_str_make_new(const mp_obj_type_t *type_in, size_t n_args, size_t
 mp_obj_t mp_obj_str_split(size_t n_args, const mp_obj_t *args);
-mp_obj_t mp_obj_new_str_copy(const mp_obj_type_t *type, const byte *data, size_t len);
-mp_obj_t mp_obj_new_str_of_type(const mp_obj_type_t *type, const byte *data, size_t len);
+mp_obj_t mp_obj_new_str_copy(const mp_obj_type_t *type, const byte *data, size_t len); // for type=str, must be known to be utf-8 safe
```
"input data must be valid utf-8"

dpgeorge commented:
In `mp_obj_str_make_new` there are two calls to `utf8_check`.  I think the second can be removed, and I think the first needs to stay.

jimmo commented:
> In `mp_obj_str_make_new` there are two calls to `utf8_check`. I think the second can be removed, and I think the first needs to stay.

Yes good catch. Updated.

Now at -80 bytes.

dpgeorge commented:
Thanks for updating (and I also updated the code comment as per above).  Rebased and merged in 8a0ee5a5c04e83f04d1c62029ac5ba7c74856507 through bd4e45fd68e20af3f965b403e643fa5f71aa1b08

dpgeorge commented:
I also run the performance benchmarks before/after this PR, on PYBv1.0.  It was basically unchanged (as expected because there aren't any tests for string operations).
